lms/djangoapps/program_enrollments/rest_api/v1/tests/test_views.py

```diff
@@ -1202,6 +1203,172 @@ def setup_change_test_data(self, initial_statuses):
+        mock_user = defaultdict(lambda: None)
+        if existing_user:
+            mock_user = {self.external_user_key: self.user}
+        with mock.patch(
+            _get_users_patch_path,
+            autospec=True,
+            return_value=mock_user,
+        ):
+            response = getattr(self.client, method)(
+                url,
+                json.dumps(write_data),
+                content_type='application/json'
+            )
+            return response
+
+    def write_program_course_enrollment(
+        self,
+        method,
+        program_uuid,
+        course_id,
+        enrollment_status
+    ):
+        """ Create or update the program course enrollment through API """
+        course_post_data = [{
+            'student_key': self.external_user_key,
+            'status': enrollment_status
+        }]
+        course_url = self.get_program_course_url(program_uuid, course_id)
+        response = getattr(self.client, method)(
+            course_url,
+            json.dumps(course_post_data),
+            content_type='application/json'
+        )
+        return response
+
+    def link_user_social_auth(self):
+        """ Create the UserSocialAuth record to trigger the linkage django signal """
+        SAMLProviderConfigFactory(
+            organization=self.lms_org,
+            slug=self.organization_key
+        )
+        UserSocialAuth.objects.create(
+            user=self.user,
+            uid='{0}:{1}'.format(self.organization_key, self.external_user_key),
+            provider=self.organization_key
+        )
+
+    @ddt.data(True, False)
+    def test_enrollment_in_same_course_multi_program(self, existing_user):
+        response = self.write_program_enrollment(
+            'post', self.program_uuid, self.curriculum_uuid, 'enrolled', existing_user
+        )
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        response = self.write_program_course_enrollment(
+            'post', self.program_uuid, self.course_id, 'active'
+        )
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+        response = self.write_program_enrollment(
+            'put', self.program_uuid, self.curriculum_uuid, 'canceled', existing_user
+        )
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        response = self.write_program_course_enrollment(
+            'put', self.program_uuid, self.course_id, 'inactive'
+        )
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+        response = self.write_program_enrollment(
+            'post', self.another_program_uuid, self.another_curriculum_uuid, 'enrolled', existing_user
+        )
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        response = self.write_program_course_enrollment(
+            'post', self.another_program_uuid, self.course_id, 'active')
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+        if not existing_user:
```

Comment:
Big ol' Nit:You could probably make failure of this test more comprehensible by unrolling the ddt as separate functions with more descriptive names.

The not existing_user case corresponds to calling a helper method and then doing this code, the existing_user case corresponds to just calling the helper method.

Names could be `test_linking_account_with_multiple_course_enrollments_to_same_course` & `test_enrolling_to_single_course_by_successive_programs_while_account_linked` or something

Reply:
I know what you are saying. I had them separate before. But I found myself repeating a lot of the setup in the beginning. There are multiple ways to structure this and I found this to be the easiest for me to understand. Will leave it here for now.